### PR TITLE
DbLedgerStorage -- ReadCache

### PR DESCRIPTION
**Pull request by merlimat**
Adds a `ReadCache` class to be used from LedgerStorage. 

The read cache is used when doing read-ahead caching after reading one entry. 

The idea is that entries for the same ledger as stored in order (for each flush cycle, eg: 1min). When reading 1 entry, we expect the client to read all subsequent entries. We can then skip the RocksDb `get()` to fetch the index of the entries, by just keep reading from the entryLog, and exploiting page cache and seqeuential reads. The entries are then added to this ReadCache, where they will looked when the request for next entry comes in. 

This implementation is based on dividing the memory into multiple segments. Whenever the read cache is full, the oldest segment gets recycled and overwritten with new entries.

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadCache.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+ */
+public class ReadCache implements Closeable {
+
+    private static final int DEFAULT_MAX_SEGMENT_SIZE = 1 * 1024 * 1024 * 1024;
+
+    private final List<ByteBuf> cacheSegments;
+    private final List<ConcurrentLongLongPairHashMap> cacheIndexes;
+
+    private int currentSegmentIdx;
+    private AtomicInteger currentSegmentOffset = new AtomicInteger(0);
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadCache.java]: nit: final

**Comment by merlimat** [comment]
> one question: I think this ReadCache is general can be shared with different ledger storage implementations, right? 

Yes, same for `WriteCache` or even for `KeyValueStorage` if that's needed in any other component.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadCache.java (hunk 1 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadCache.java]: yes, I thought it was being reassigned but it's not.

**Comment by sijie** [comment]
@merlimat we can consider sharing the components with other implementations once the db ledger storage change is merged.

**Comment by sijie** [comment]
merging this now to unblock subsequent db ledger storage changes.

**Pull request closed without merging**